Public/Src/Tools/DropDaemon/DropDaemon.cs

```diff
@@ -567,43 +569,73 @@ private async Task ReportStatisticsAsync()
             }
         }
 
+        private delegate TResult ErrorFactory<TResult>(string message, IpcResultStatus status);
+
         private static Task<IIpcResult> WrapDropErrorsIntoIpcResult(Func<Task<IIpcResult>> factory)
         {
-            return HandleKnownErrors(
+            return HandleKnownErrorsAsync(
                 factory,
-                (errorMessage) => new IpcResult(IpcResultStatus.ExecutionError, errorMessage));
+                (errorMessage, status) => new IpcResult(status, errorMessage));
         }
 
         private static Task<TDropEvent> WrapDropErrorsIntoDropEtwEvent<TDropEvent>(Func<Task<TDropEvent>> factory) where TDropEvent : DropOperationBaseEvent
         {
-            return HandleKnownErrors(
+            return HandleKnownErrorsAsync(
                 factory,
-                (errorMessage) =>
+                (errorMessage, errorKind) =>
                 {
                     var dropEvent = Activator.CreateInstance<TDropEvent>();
                     dropEvent.Succeeded = false;
                     dropEvent.ErrorMessage = errorMessage;
+                    dropEvent.AdditionalInformation = RenderIpcStatus(errorKind);
                     return dropEvent;
                 });
         }
 
-        private static async Task<TResult> HandleKnownErrors<TResult>(Func<Task<TResult>> factory, Func<string, TResult> errorValueFactory)
+        private static string RenderIpcStatus(IpcResultStatus status)
+        {
+            return status.ToString();
+        }
+
+        private static IpcResultStatus ParseIpcStatus(string statusString, IpcResultStatus defaultValue = IpcResultStatus.ExecutionError)
+        {
+            return Enum

```

Comment:
>return Enum [](start = 12, length = 11)

I think calling Enum.TryParse() would look a little bit cleaner. Is there a benefit of doing the parsing/casting this way?

Reply:
you are right, will change